From 2083ed9b84781cb74989d484f041b56cd62a69dc Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 12 Aug 2021 17:31:29 +0300
Subject: [PATCH] [SW-0] gaudi: security: improve lbw handover

To improve the BTL settings replacement w/o creating an opportunity
window, closing the cfg space access from pcie before reconfiguring
the rest of the range registers

Change-Id: Id7a380e699c1316fcb2525b085e5750e8765f23c
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/arch/gaudi/security.c b/arch/gaudi/security.c
index 8d21ae4..96eb51f 100644
--- a/arch/gaudi/security.c
+++ b/arch/gaudi/security.c
@@ -1419,6 +1419,7 @@
 	LBW_RR_DBG_ETR,
 };
 
+#define HOST_PHYS_BASE		0x8000000000ull
 static void preboot_secure_lbw(void)
 {
 	u64 min, max;
@@ -1428,6 +1429,28 @@
 	 * meaning - not inclusive, so we need to provide it the min address
 	 * below the forbidden one and max address above the forbidden one.
 	 */
+
+	/*
+	 * closing the cfg space before everything else in order to avoid
+	 * creating a security hole during the configuration of the range regs
+	 */
+	/* CFG LOW RANGE: RR(PCIE_DYN_RR_CFG_LOW) */
+	min = HL_CFG_BASE - 1;
+	max = mmPSOC_GLOBAL_CONF_BASE;
+	preboot_secure_pcie_rr(PCIE_DYN_RR_CFG_LOW, min, max);
+
+	/* CFG HIGH RANGE: RR(PCIE_DYN_RR_CFG_HIGH) */
+	min = mmPSOC_GPIO0_BASE - 1;
+	/*
+	 * The PCIe RR can take only values of 32 bits, then H/W is
+	 * extending these with 0x7F in the bits 32-39.
+	 * This means we cannot provide the HOST_PHYS_BASE as it will
+	 * just turn into 0x7F00000000 which is not what we want.
+	 * Subtract 1 from HOST_PHYS_BASE to hae a valid value.
+	 */
+	max = HOST_PHYS_BASE - 1;
+	preboot_secure_pcie_rr(PCIE_DYN_RR_CFG_HIGH, min, max);
+
 	/* MME1_QM: RR(LBW_RR_MME1_QM) */
 	min = mmMME1_QM_BASE - 1;
 	max = mmMME2_ACC_BASE;
@@ -1783,7 +1806,6 @@
 
 #define SRAM_SIZE	0x1400000	/* 20MB */
 #define SRAM_END	(SDRAM_BASE + SRAM_SIZE)
-#define HOST_PHYS_BASE	0x8000000000ull
 
 enum preboot_sec_rr_names {
 	SRAM_BGN_TO_VERS = 0,
